WIP: RFC: Wasmtime 1.0 (#2)
alexcrichton/rfcs-2: accepted/wasmtime-one-dot-oh.md

expected from the stable release of a project.

On a spectrum of "Wasmtime never breaks any API ever" to "Wasmtime breaks all
its APIs every release" Wasmtime will reside towards the first half here. In
other words Wasmtime developers will be expected to rarely and deliberately
break existing APIs, but will also be allowed to break APIs as necessary. The
major goal here is "hassle-free upgrades" inspired by the Rust release process
where API stability in the absolute literal sense is deterimental to the
development of the project, but deliberate effort and work is done to ensure
upgrades are as seamless as possible.

Users should expect that they will regularly need to increase the major version
of Wasmtime to upgrade Wasmtime usage in their projects. Furthermore users
should expect that between upgrades they may need to tweak some APIs and update
them to continue to use the latest-and-greatest Wasmtime. At the same time
though users should expect that they will not be blindsided by major
refactorings. For example changes like [Wasmtime's new API][new-api] will still
go through the RFC process and will be deliberately scheduled and widely

> advertised before they're released.

Reviewer: Maybe worth explicitly mentioning that, even though they will go through the RFC process when we need them, we expect such large changes to be rare.

have the same version as the `wasmtime` Rust crate, and each release they'll be
able to make API-breaking changes. As noted in the previous section, though,
each release is expected to be a relatively hassle-free upgrade, so while
API-breaking changes are allowed they're not necessarily encouraged.

Wasmtime will continue to create a tag for all released versions of Wasmtime
with a corresponding GitHub release as-is done today for the CLI and C API
binaries. Embeddings will all be tagged and released as appropriate to
language-specific package managers (such as crates.io and PyPI). Wasmtime will
always be released from the `main` branch of the Wasmtime repository itself,
which means that all development is happening on `main` and once something lands
it's guaranteed to be in the next release. Note that bug fixes and such for
historical releases is discussed later in this proposal.

Releasing a new version of Wasmtime every 4 weeks can be quite rapid for some
users who don't necessarily want to stay up-to-date with the latest and greatest
of Wasmtime, but still want the stability of a production-ready WebAssembly
runtime. For these users, this leads well into the next section ...

> ### Extended Support Releases

Reviewer: +1 for renaming this "Long-Term Support (LTS) Releases"

  fixes to the previous stable release of Wasmtime, but not others. This was
  done in an ad-hoc and not well-defined manner, and isn't the best way to do
  this moving forward.

The motivation for this RFC is to effectively fix all these issues. Wasmtime is,
as a codebase, technically ready for use in production. This RFC intends to
cover the rest of the ground necessary to make Wasmtime suitable for production
use, mostly on process-aspects and such. Once this RFC is implemented we, as
project maintainers, and users, as developers using Wasmtime, should all be
confident that Wasmtime feels "1.0" and is production ready in all aspects.

Many projects in the open source ecosystem are "production ready" and can serve
as inspiration to draw from for each of the above points. Release processes for
various projects can widely differ and serve different goals for both
maintainers and users. What's appropriate for one project doesn't necessarily
mean that it's also appropriate for Wasmtime itself. This proposal upholds a
number of goals that are specific to Wasmtime itself and how it's expected
Wasmtime will be used:

> * Wasmtime should be released to users in a predictable fashion. For exmaple

Reviewer 2: ```suggestion
* Wasmtime should be released to users in a predictable fashion. For example
```

  between Wasmtime versions. For example "how do I get the latest version?" or
  "how do I upgrade from my current version?" should be easy to answer.

* Developers of Wasmtime should not be hindered by processes to land
  improvements to Wasmtime. For example "when can I land my 10% perf increase
  that tweaks APIs" should be easy to answer, and the answer shouldn't be
  "six months from now". Furthermore bugfixes which may break APIs should also
  not be difficult to land.

* Users should be able to clearly understand what version of Wasmtime they're
  using. It's possible to use Wasmtime through a variety of means such as the
  numerous embedding APIs for each language, and it should always be clear what
  version of the project you're using.

The constraints here do not map perfectly to what other projects do all the
time. For example Wasmtime will not have as strong of a committment to API
stability as the Rust language and standard library. Wasmtime does want,
however, to maintain the "largely hassle-free upgrade" experience inspired by
Rust's example, though. This proposal attempts to balance these various concerns

> in such a way that is most beneficial for the Wasmtime project itself.

Reviewer 2: This sounds a little bit as though the balance aimed at is one that's beneficial to maintainers over other concerns. Perhaps something like this might work?
```suggestion
Rust's example, though. This proposal attempts to identify the right balance 
between these various concerns for the Wasmtime project.
```

is ever-changing and ever-improving where there are almost always active
proposals in flight to extend and modify WebAssembly. The nature of Wasmtime and
its goals mean that **Wasmtime's stability story will not include API stability
in the literal sense of "Wasmtime will always be 1.0 and never 2.0"**.

For Rust users this may feel a bit discombobulating where Rust is likely to
always be "1.x", but this RFC is proposing that Wasmtime will likely never
remain indefinitely "1.x". Instead users and developers can expect that Wasmtime
will regularly release new major versions of its tooling. With the expectation
that Wasmtime will release new major versions it enables developers of Wasmtime
itself to land API-breaking changes. This makes it much easier to implement new
WebAssembly proposals, improve the runtime, fix bugs, make APIs more ergonomic,
etc. These are seen as critical abilities needed to improve Wasmtime over time.
As mentioned in the motivation section, Wasmtime developers themselves should
not feel overly burdened or having to contort their code to land changes which
are seen as important. Allowing breaking changes is seen as key to enabling

> this.

Reviewer 2: While you expand on the reasons for this approach further down, people will form an opinion based on what they read here, so it might be better to lead with "we have to break API compat because wasm features", e.g. like this:
```suggestion
For Rust users this may feel a bit discombobulating where Rust is likely to
always be "1.x". This isn't viable for Wasmtime: the addition of new WebAssembly 
features will often require breaking API compatibility, so we don't have much 
choice in the matter. The ability to break API compatibility also enables
improvements to the runtime that wouldn't be possible without changes to the 
embedding API. These are seen as critical abilities needed to improve Wasmtime 
over time. As mentioned in the motivation section, Wasmtime developers themselves 
should not feel overly burdened or having to contort their code to land changes 
which are seen as important. Allowing breaking changes is seen as key to enabling
this.
```

Naturally, though, the world is not entirely made up of Wasmtime developers.
Instead it's much more likely that the number of users of Wasmtime far
outweighs the number of developers on Wasmtime. A stream of constant breaking
changes makes it feel harder to use Wasmtime since it can feel more difficult to
keep up with the latest and receive bug fixes and new features. Wasmtime's
stability story, consequently, will not be on the end of the spectrum where
breaking changes are landed willy-nilly. This means that there's a tension set
up between developers of Wasmtime and users of Wasmtime, but this is to be
expected from the stable release of a project.

On a spectrum of "Wasmtime never breaks any API ever" to "Wasmtime breaks all
its APIs every release" Wasmtime will reside towards the first half here. In
other words Wasmtime developers will be expected to rarely and deliberately
break existing APIs, but will also be allowed to break APIs as necessary. The
major goal here is "hassle-free upgrades" inspired by the Rust release process

> where API stability in the absolute literal sense is deterimental to the

Reviewer 2: ```suggestion
where API stability in the absolute literal sense is detrimental to the
```

development of the project, but deliberate effort and work is done to ensure
upgrades are as seamless as possible.

Users should expect that they will regularly need to increase the major version
of Wasmtime to upgrade Wasmtime usage in their projects. Furthermore users
should expect that between upgrades they may need to tweak some APIs and update
them to continue to use the latest-and-greatest Wasmtime. At the same time
though users should expect that they will not be blindsided by major
refactorings. For example changes like [Wasmtime's new API][new-api] will still
go through the RFC process and will be deliberately scheduled and widely
advertised before they're released. Note that it is expected that changes of

> that magnitude are expected to be quite rare.

Reviewer 2: ```suggestion
advertised before they're released. Note also that changes of that magnitude are
expected to be quite rare.
```

* The `wasmtime` Rust crate
* The `wasmtime-*` embedding APIs
* The Wasmtime C API

Notably this proposal for stability does not include Cranelift or `wasm-tools`.
These projects are expected to have their own story for stability. Release will
be made for these crates to accomodate Wasmtime itself, but the release process
here will not automatically apply to these crates otherwise. For example
Cranelift will reach 1.0 on its own cadence and otherwise will continue to
receive major version bumps as Wasmtime is released. The `wasm-tools` crates
will continue to be released on-demand for consumers (including Wasmtime) and
have their own stability (such as `wat` being very stable but `wast` being much
less API-stable).

Additionally this stability proposal does not, at this time, include the
`wasmtime-wasi` or `wasi-common` crates. The `wasmtime-wasi` crate will
continue to be "production ready" in the sense that they will be promoted to the
same version number as Wasmtime itself, but they will not carry the same promise
of general API stability that the Wasmtime crate will carry. These crates are
likely to receive more breaking changes until at such a point in the future that

> they have become more stable (likely through another RFC).

Reviewer 2: I think this makes sense for now, yes. I wonder if eventually we should simply fold `wasmtime-wasi` into the `wasmtime` crate itself, though: we'd expect basically all uses of Wasmtime to use at least some WASI functionality. But that's something we'll be able to figure out eventually, in the meantime it's good to have this escape hatch to avoid having to bump `wasmtime`'s major version quite as frequently.

follow the rest of the release process and procedures defined here.

## What does it mean for a feature to be stable?

Features enabled-by-default and implemented in Wasmtime are expected to clear at
least a minimum threshold of stability, including:

* The feature must be throughly tested in Wasmtime's CI on major platforms (at
  this time x86\_64 and AArch64)

* If the feature is for a WebAssembly upstream proposal, all spec tests must be
  enabled and passing and the proposal must be in [stage 4 or
  later](https://github.com/WebAssembly/meetings/blob/main/process/phases.md).

* The feature must have no open questions, design concerns, or serious known
  bugs.

* The feature must have support in fuzzers and have been fuzzed for at least a
  week. Additionally we should be confident that the fuzzing support for this
  feature is exercising the necessary bits in Wasmtime thoroughly.

* The feature is supported in the Rust API, the C API, and at least one other
  language's embedding API.

* A member of the Bytecode Alliance must be "on the hook" for maintenance of
  this feature.

Features can be implemented in-tree even if they do not meet these criteria but

> the features must either be disabled-by-default at either runtime or compile

Reviewer 2: ```suggestion
the features must be disabled-by-default at either runtime or compile
```

time. If a feature's interim implementation does not have an undue compile-time
or runtime footprint then it can be off-by-default at runtime but compiled in by
deafult. If, however, an in-progress feature has a significant compile-time or
runtime footprint it must be disabled at compile-time by default.

Features implemented in-tree but not currently stabilized are also subject to
removal if there is no active progress being made on the feature. For example if
this RFC is approved **the `lightbeam` feature of Wasmtime would be removed**.

> **Note**: This RFC proposes removing `lightbeam` purely because it is not
> production quality (it is disabled by default at compile time) and has not
> been maintained for quite some time. The Wasmtime project is still interested
> in pursuing alternative backends (e.g. baseline compilers), although this is

> > not being actively pursued by anyone at this time.

Reviewer 2: ```suggestion
> **Note**: This RFC proposes removing `lightbeam` because it isn't actively 
> maintained, and as a result hasn't been compiling  successfully for quite 
> some time. The Wasmtime project is still interested in the addition of a
> baseline compiler, but the addition of one would be proposed in an RFC and
> undergo careful evaluation to ensure the design aligns with Wasmtime's
> goals and requirements, and the compiler has a solid maintenance story.
```

## Release Process and Cadence

Wasmtime intends to follow in the footsteps of many other projects on the matter
of release cadence with a frequent and predictable release cycle. **This RFC

> proposes releasing Wasmtime on Tuesday every 4 weeks**.

Reviewer 2: I guess this is something to discuss with the wider group, but I wonder if a release every 4 weeks is a better schedule than, say, Rust's 6 weeks. I guess to some degree that depends on how much work a release ends up being. (I do see how this has nice properties wrt LTS release schedules, etc, though 😄)

Author: Yeah I think we'll determine, uh, real fast whether this is reasonable or not. I also hope to invest time in to automating things further beyond where they are today, ideally with the process looking like:

* A bot sends a PR with the version bump
* We manually merge PR at the desired time (and apply fixes as necessary)
* Thats... it?

So ideally all we do to make a release is hit the "yes as a human I confirm this should happen" button.

The goal is to publish work and improvements in Wasmtime on a relatively rapid
schedule to ensure that the latest-and-greatest is available for usage. This is
also intended to signal to users that consumers of Wasmtime should expect to
stay up-to-date with Wasmtime and it's not appropriate to ossify a version for
years and expect it to continue to work and receive everlasting maintenance.

Each release of Wasmtime will bump the major version number of Wasmtime itself.
This means, for example, that Wasmtime 2.0 will be released one month after
Wasmtime 1.0. After one year Wasmtime will be at 12.0. At this time it is not
planned that "minor" releasese will be made of the 1.1.0 variety. This major
version bump on each release allows all of Wasmtime's embeddings and
sub-projects to stay in sync, for example the `wasmtime-py` Python bindings will
have the same version as the `wasmtime` Rust crate, and each release they'll be
able to make API-breaking changes. As noted in the previous section, though,
each release is expected to be a relatively hassle-free upgrade, so while

> API-breaking changes are allowed they're not necessarily encouraged.

Reviewer 2: I reworded this a bit to more thoroughly explain why the cross-language version number alignment favors regular major version bumps:
```suggestion
Each release of Wasmtime will bump the major version number of Wasmtime itself.
This means, for example, that Wasmtime 2.0 will be released one month after
Wasmtime 1.0. After one year Wasmtime will be at 12.0. At this time it is not
planned that "minor" releasese will be made of the 1.1.0 variety. Note that this
means means that not all major releases will actually contain breaking changes.
The reason we propose to nevertheless always bump the major version is that this
allows us to keep Wasmtime's version number in sync with that of its various
language-specific embeddings: as a developer using Wasmtime, you shouldn't have
to worry about how how the version number of your language's Wasmtime embedding
lines up with that of Wasmtime itself. E.g., if you use`wasmtime-py 7.0`, you
can be sure that you're using Wasmtime 7.0. Trying to keep major version bumps
to a minimum while keeping version numbers aligned would force us to still bump
Wasmtime's and all embeddings' version numbers whenever there's a breaking 
change in even a single language-specific embedding. As noted in the previous 
section, though, each release is expected to be a relatively hassle-free 
upgrade, so while API-breaking changes are allowed they're not necessarily 
encouraged.
```

Wasmtime will continue to create a tag for all released versions of Wasmtime

> with a corresponding GitHub release as-is done today for the CLI and C API

Reviewer 2: ```suggestion
with a corresponding GitHub release as is done today for the CLI and C API
```

binaries. Embeddings will all be tagged and released as appropriate to
language-specific package managers (such as crates.io and PyPI). Wasmtime will
always be released from the `main` branch of the Wasmtime repository itself,
which means that all development is happening on `main` and once something lands
it's guaranteed to be in the next release. Note that bug fixes and such for

> historical releases is discussed later in this proposal.

Reviewer 2: ```suggestion
historical releases are discussed later in this proposal.
```

LTS releases. **Wasmtime will support two active LTS versions at any point in
time, each supported for 10 months at a time**.

For users who want to use Wasmtime but are not interested in upgrading monthly,
this will be a suitable alternative where Wasmtime LTS versions will have the
following properties while the version is considered "supported":

* They will receive security fixes.
* They will receive bug fixes related to correctness.
* They will _not_ receive new features.

For example if a new WebAssembly proposal is implemented it will not be
backported to LTS versions. Additionally if a bug is fixed in a wasm proposal
that is off-by-default in an LTS version it will also not be backported. Other
bug fixes (and of course security fixes) will be backported to supported LTS
versions.

An LTS version is considered "supported" for 10 release cycles, or 40 weeks (~10
months). Wasmtime will have two LTS versions at any point in time, with a new
LTS happening every 5 releases (20 weeks, ~5 months). For ease of remembering
what's an LTS and what isn't, all releases of Wasmtime divisible by 5 will be
LTS version, with the exception of 1.0 being an LTS version as well. For example

> the LTS versions of Wasmtime will be 1.0, 5.0, 10.0, 15.0, ...

Reviewer 2: Might be good to include something like "see below for details on the ongoing versioning of LTS releases"

Author: Do you mean in relation to the patch number? Otherwise I'm not sure what info you're alluding to?

This cadence means that users who do not want to upgrade Wasmtime monthly will
be expected to upgrade Wasmtime at least every 10 months, likely every 5 months.
These upgrades are likely to be less "hassle-free" than each individual version
upgrade since it will accumulate at least 5 releases worth of minor breaking

> changes.

Reviewer 2: ```suggestion
changes. But in exchange, users of LTS versions have a 5-months window to upgrade while staying on a supported version.
```

LTS versions will be maintained in separate branches of the Wasmtime repository.
At this time it's expected that LTS versions will not needed separate branches
in separate embedding repos and tags will suffice. The branch names for the

> Wasmtime repository will be `esr-latest` and `esr-oldest` for the most recent

Reviewer 2: ```suggestion
Wasmtime repository will be `lts-latest` and `lts-oldest` for the most recent
```

and the second-most-recent LTS version.

## Backports - Security fixes

With an established concept of releases and LTS for Wasmtime this provides a

> framework to discuss how security issues are handled in Wasmtime. **Security

Reviewer 2: We already have a process for handling security issues, so might be worth clarifying that this is only about their handling within the release process.
```suggestion
framework to discuss how security issues are handled in Wasmtime's release process. **Security
```

issues will be applied to the current version and supported LTS versions of
Wasmtime**.  Security issues will always be released as patch releases. The
current version of Wasmtime at the time of the issue being made public will be
patched in addition to the LTS versions at the time.

For example, if Wasmtime is currently at 12.0 then the current LTS versions are
5.0 and 10.0. If a security issue is identified at this time then the following
new releases will be made available: 5.0.1, 10.0.1, 12.0.1. No other versions of

> Wasmtime will be patched, for example 11.0 will not be patched as it's neither

Reviewer 2: I wonder if we should include some provision for still updating the previous version if a patch becomes necessary early in the cycle? Otherwise we all of a sudden force people to update within, say, a few days instead of 4 weeks.

This might even point towards regularly doing patch releases for the latest and the previous versions, though that might be more work than we're willing to commit to.

Author: Yeah I'll add some words that these are the minimum versions that we patch, but we can always do more.

    mean that the embedding APIs also didn't have any breaking changes (or vice
    versa too).

  By having a new major version every 4 weeks it's definitely a form of "least
  common denominator" solution but it also provides a predictable versioning
  number as well as a clear indication of what is LTS and what isn't. For these
  reasons it's proposed here to use a new major version every 4 weeks instead of
  gauging via breaking changes or not.

# Open questions
[open-questions]: #open-questions

- What sort of automation is required for Wasmtime developers to keep up with
  once-a-month releases? The process described here is relatively expansive and
  can take quite some effort to keep up with. For example:

  * Can we entirely automate releases?
  * Could this automation also get connected to all the other embeddings and
    repositories?
  * How hard will it be to make a patch release with a bug fix?
  * How hard will it be to remember what branches need backporting?

- Are releases skipped/delayed when they correspond to work holidays? For
  example if a release happens to coincide with Christmas should it be delayed?

>   Skipped? Perhaps with entirely automated releases this wouldn't matter!

Reviewer 2: I think we might answer this in this RFC already. Specifically, I'd propose including language in the relevant section above stating that a release might happen a week early or late to avoid coinciding with public holidays, but that the next release will still happen at its expected date.

tiers. This tier does not come with many guarantees associated with it, and is
intended to be a breeding ground and call-for-help for components to reach tier
2 or higher status.

Components in tier 3 may still meet various requirements of tier 1 and 2 while
still being classified as tier 3. For example components here may not have code
review but could still be of a high-enough quality to use in production.
Similarly they might be released at the same cadence of Wasmtime itself but
lacking in API stability.

Today this tier includes:

* `wasmtime-py` - Python bindings for Wasmtime
* `wasmtime-go` - Go bindings for Wasmtime
* `wasmtime-dotnet` - .NET bindings for Wasmtime
* `wasmtime-cpp` - C++ bindings for Wasmtime
* `witx-bindgen` - Canonical API bindings generator for Interface Types

Each of these projects is missing at least one criteria from the list of Tier 2
requirements. As is inherent to this tier, however, this is a call-for-support
for others who are interested in moving these projects up a tier of support. For
example most of these projects primarily need more maintainers to help review
bindings for language idioms and implementation, and that's all that's necessary

> to move into tier 2.

Reviewer 2: ```suggestion
to move into tier 2.

It's also worth highlighting that the intent here is to establish a deliberately high bar of quality we commit to. In many established open source projects, the level of quality, and the development and reviewing practices some or all of these components have and use would be deemed "good enough". And in many domains they might be. We're acutely aware of Wasmtime's, and more generally WebAssembly's use in mission- and security-critical environments, and we believe that we need to hold ourselves to higher standards accordingly.
```
